### PR TITLE
Adds /supportMatrix and /supportMatrix/servers resources to REST API

### LINKED ISSUE #163: REST: Consider a call that returns All MP versions, Servers and APIs in one hit
@yeekangc suggested we have a call that returns all mapping MPVersion:Servers:APIs for the usage in tools such as VSCode. The current API forces you to do this in 2 steps, first fetch list of MPs and and the second selects particular MP version:
```
$ curl https://start.microprofile.io/api/1/mpVersion

["MP22","MP21","MP20","MP14","MP13","MP12"]
```
```
$ curl https://start.microprofile.io/api/1/mpVersion/MP14
{
  "supportedServers": [
    "PAYARA_MICRO",
    "LIBERTY",
    "TOMEE",
    "KUMULUZEE"
  ],
  "specs": [
    "CONFIG",
    "FAULT_TOLERANCE",
    "JWT_AUTH",
    "HEALTH_METRICS",
    "HEALTH_CHECKS",
    "OPEN_API",
    "OPEN_TRACING",
    "REST_CLIENT"
  ]
}
```

Opened for consideration.

### PR DESCRIPTION
**Pull request by Karm**
Fixes #163

See the [documentation](https://github.com/Karm/microprofile-starter/blob/issue-163/src/main/resources/REST-README.md#integration-with-ides) of the new API resource.

I would like to ask @yeekangc to review if it helps with IDE integration.

It can be tested live on:

```
$ curl  -i https://test-start.microprofile.io/api/supportMatrix/servers
```
and
```
$ curl  -i https://test-start.microprofile.io/api/supportMatrix
```

Signed-off-by: Michal Karm Babacek <karm@fedoraproject.org>

### CODE REVIEW

**Comment by Karm** [comment]
@Karm Update Swagger :)

**Comment by yeekangc** [comment]
Looks nice, @Karm.

The one question as was briefly discussed is when Java 11 comes, what will it be.  It can be a separate item to consider something like a `/supportMatrix/java` then.

File: src/main/resources/REST-README.md [hunk 1]
```diff
@@ -244,3 +244,153 @@ PS C:\> $headers = @{
...
 PS C:\> Invoke-WebRequest -InFile ./all.json -OutFile project.zip -Method Post -Uri "https://start.microprofile.io/api/1/project" -Headers $headers
 ```
+
+# Integration with IDEs
+
+If it is preferred to acquire all valid options in a one request the ```supportMatrix``` call is recommended.
+The integration code [SHOULD](https://tools.ietf.org/html/rfc2119) use [Etag](https://tools.ietf.org/html/rfc7232#section-2.3) response header
+and [If-None-Match](https://tools.ietf.org/html/rfc7232#section-3.2) request header so as to avoid unnecessary deserialization of identical responses.
+
+## Microprofile versions as keys
```
Comment (yeekangc) [inline src/main/resources/REST-README.md]:
```suggestion
## MicroProfile versions as keys
```

Comment (Karm) [reply src/main/resources/REST-README.md]: Thx. Addressed.

**Comment by Karm** [comment]
@yeekangc 

> The one question as was briefly discussed is when Java 11 comes,
> what will it be. It can be a separate item to consider
> something like a /supportMatrix/java then.

I think so. Let's add such resource when it becomes pertinent.

**Comment by Karm** [comment]
@Emily-Jiang @rdebusscher 

Any comments? Can we merge and do the 9.3 release live?

File: src/main/java/org/eclipse/microprofile/starter/rest/APIService.java [hunk 2]
```diff
@@ -78,18 +82,38 @@
...
+                                                .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()),
+                                        Stream.of(MicroprofileSpec.values())
+                                                .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()))));
 
                     }
 
                 };
+        mpvToOptionsEtag = new EntityTag(Integer.toHexString(31 * version.getGit().hashCode() + mpvToOptions.hashCode()));
+        // Keys are servers and values are MP versions and specs
+        serversToOptions =
```
Comment (rdebusscher) [inline src/main/java/org/eclipse/microprofile/starter/rest/APIService.java]: Same remark as for mpvToOptions -> unreadble!

This is equivalent (this does not include the random order, see other comment)

        serversToOptions = new HashMap<>();
        for (SupportedServer supportedServer : SupportedServer.values()) {
            Map<MicroProfileVersion, List<MicroprofileSpec>> serverOption = new HashMap<>();
            for (MicroProfileVersion mpVersion : supportedServer.getMpVersions()) {
                List<MicroprofileSpec> specList = new ArrayList<>();
                for (MicroprofileSpec spec : MicroprofileSpec.values()) {
                    if (spec.getMpVersions().contains(mpVersion)) {
                        specList.add(spec);
                    }
                }
                serverOption.put(mpVersion, specList);

            }
            serversToOptions.put(supportedServer, serverOption);
        }

or with a little bit of Stream ...

        serversToOptions = new HashMap<>();
        for (SupportedServer supportedServer : SupportedServer.values()) {
            Map<MicroProfileVersion, List<MicroprofileSpec>> serverOption = new HashMap<>();
            for (MicroProfileVersion mpVersion : supportedServer.getMpVersions()) {
                List<MicroprofileSpec> specList =
                        Arrays.stream(MicroprofileSpec.values())
                                .filter(ms -> ms.getMpVersions().contains(mpVersion))
                                .collect(Collectors.toList());
                serverOption.put(mpVersion, specList);

            }
            serversToOptions.put(supportedServer, serverOption);
        }

File: src/main/java/org/eclipse/microprofile/starter/rest/APIService.java [hunk 3]
```diff
@@ -78,18 +82,38 @@
...
+        // Keys are MP versions and values are servers and specs
         mpvToOptions =
                 new HashMap<MicroProfileVersion, MPOptionsAvailable>(MicroProfileVersion.values().length) {
                     {
-                        Stream.of(MicroProfileVersion.values()).forEach(mpv -> put(mpv, new MPOptionsAvailable(
-                                Stream.of(SupportedServer.values())
-                                        .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()),
-                                Stream.of(MicroprofileSpec.values())
-                                        .filter(v -> v.getMpVersions().contains(mpv)).collect(Collectors.toList()))));
+                        Stream.of(MicroProfileVersion.values()).filter(mpv -> mpv != MicroProfileVersion.NONE)
```
[the remark this answers is not in the document]
Comment (Karm) [reply src/main/java/org/eclipse/microprofile/starter/rest/APIService.java]: @rdebusscher Thank you for the feedback. Ad anonymous inner class  - I was aware, yet I considered it as "anything goes" given the fact it is supposed to run once...
Ad nested streams - yes, I clearly went off the hook there.

It is refactored now. Please check dbb941f

File: src/main/java/org/eclipse/microprofile/starter/rest/APIService.java [hunk 4]
```diff
@@ -78,18 +82,38 @@
...
 
                     }
 
                 };
+        mpvToOptionsEtag = new EntityTag(Integer.toHexString(31 * version.getGit().hashCode() + mpvToOptions.hashCode()));
+        // Keys are servers and values are MP versions and specs
+        serversToOptions =
+                new HashMap<SupportedServer, Map<MicroProfileVersion, List<MicroprofileSpec>>>(SupportedServer.values().length) {
+                    {
+                        Stream.of(SupportedServer.values()).sorted()
```
[the remark this answers is not in the document]
Comment (Karm) [reply src/main/java/org/eclipse/microprofile/starter/rest/APIService.java]: @rdebusscher 
Fixed.

```/supportMatrix``` has order of servers on serialization randomized [here](https://github.com/Karm/microprofile-starter/blob/issue-163/src/main/java/org/eclipse/microprofile/starter/rest/model/MPOptionsAvailable.java#L37:L41)  (worked before)

```/supportMatrix/servers``` has order of servers randomized before serialization [here](https://github.com/Karm/microprofile-starter/blob/issue-163/src/main/java/org/eclipse/microprofile/starter/rest/APIService.java#L174:L182) (new change)

File: src/main/java/org/eclipse/microprofile/starter/rest/APIService.java (hunk 2 above)
Comment (Karm) [reply src/main/java/org/eclipse/microprofile/starter/rest/APIService.java]: @rdebusscher Refactored.

File: src/main/resources/REST-README.md [hunk 5]
```diff
@@ -244,3 +244,153 @@ PS C:\> $headers = @{
...
+      "TOMEE",
+      "LIBERTY",
+      "KUMULUZEE",
+      "HELIDON"
+    ],
+    "specs": [
+      "CONFIG",
+      "FAULT_TOLERANCE",
+      "JWT_AUTH",
+      "HEALTH_METRICS",
```
Comment (Emily-Jiang) [inline src/main/resources/REST-README.md]: METRICS

File: src/main/resources/REST-README.md [hunk 6]
```diff
@@ -244,3 +244,153 @@ PS C:\> $headers = @{
...
+      "KUMULUZEE",
+      "THORNTAIL_V2",
+      "LIBERTY",
+      "PAYARA_MICRO"
+    ],
+    "specs": [
+      "CONFIG",
+      "FAULT_TOLERANCE",
+      "JWT_AUTH",
+      "HEALTH_METRICS",
```
Comment (Emily-Jiang) [inline src/main/resources/REST-README.md]: METRICS

**Comment by Karm** [comment]
Merged, Rudy and Emily approved.

**Pull request merged**